src/vs/platform/instantiation/common/instantiationService.ts

```diff
@@ -246,15 +249,52 @@ export class InstantiationService implements IInstantiationService {
 			// Return a proxy object that's backed by an idle value. That
 			// strategy is to instantiate services in our idle time or when actually
 			// needed but not when injected into a consumer
+
+			// return "empty events" when the service isn't instantiated yet
+			const earlyListeners = new Map<string, LinkedList<Parameters<Event<any>>>>();
+
 			const idle = new IdleValue<any>(() => {
 				const result = child._createInstance<T>(ctor, args, _trace);
+
+				// early listeners that we kept are now being subscribed to
+				// the real service
+				for (const [key, values] of earlyListeners) {
+					const candidate = <Event<any>>(<any>result)[key];
+					if (typeof candidate === 'function') {
+						for (const listener of values) {
+							candidate.apply(result, listener);
+						}
+					}
+				}
+				earlyListeners.clear();
+
 				return result;
 			});
 			return <T>new Proxy(Object.create(null), {
 				get(target: any, key: PropertyKey): any {
+
+					if (!idle.isInitialized) {
+						// looks like an event
+						if (typeof key === 'string' && (key.startsWith('onDid') || key.startsWith('onWill'))) {
```

Comment:
Should there be a tslint rule making sure we don't call non-event properties something which might trigger this?

Reply:
We have such a rule for vscode.dts-files but for core it is much harder. In neither case (core and API) we have type information but for the API there is only a single type called `Event` so we make a very qualified guess. Also, in the API we always have type annotations which we don't always have in core. Last, not all events in core are properly named. They should be but I don't want to block this effort on that. For now it is opt-in: you want a lean and fast service, play by the book of conventions

Reply:
I just think it's going to come up in some standup in the future after a 2 hour "why is my service not running" investigation. 😄 

But yeah I do see the hassle of getting this right. And you're right, let's not block on that.

Reply:
> in the future after a 2 hour "why is my service not running" investigation

That's shouldn't happen. Worst case is today's case were listing forces service creation